### PR TITLE
Add --tag-latest to tag with the 'latest' tag or a tag without a version

### PR DESCRIPTION
**Pull request by JayH5**
My aim for the API for working with different versions is this:
#### `--tag TAG...`:
Change the tag of the image:
* `test-image:abc` -> `test-image:def`

#### `--tag-version VERSION`:
Prepend the tag with a version:
* `test-image:abc` -> `test-image:1.2.3-abc`

#### `--tag-latest`:
Also tag the image *without* a version, and tag with `latest` if the tag minus the version is empty:
* `test-image:abc` -> `[test-image:1.2.3-abc, test-image:abc]`
* `test-image` -> `[test-image:1.2.3, test-image:latest]`

#### `--tag-semver`:
Tag with all the parts of the version, like some versions of `deploy.py`:
* `test-image:abc` -> `[test-image:1.2.3-abc, test-image:1.2-abc, test-image:1-abc]`

---

The last one hasn't been implemented yet and I'm not 100% sure about the name. The idea is that all these options can be used together and it'll cover all the use cases of `deploy.py`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,32 @@
...
+A command-line tool to help generate tags and push Docker images to a registry. Simplifies deployment of Docker images from CI services such as Travis CI.
+
+In a single command, `docker-ci-deploy` can:
+* Change the tags on images
+* Add version information to image tags
+* Add registry addresses to image tags
+* Login to a registry
+* Push tags to a registry
+
+The best way to try out `docker-ci-deploy` is to give it a spin with the `--dry-run` flag and observe all the :
```
Comment (jerith) [inline README.md]: Observe all the?

File: README.md [hunk 2]
```diff
@@ -4,7 +4,32 @@
...
+```
+> $ docker-ci-deploy --tag-version 2.7.13 --tag-latest \
+      --registry registry:5000 --login 'janedoe:pa55word' --dry-run \
```
Comment (jerith) [inline README.md]: Should this be `pa$$word` to match other places in the README?

(I know we switched to `pa55word` in tests to make matching easier, but that's not really relevant here.)

File: docker_ci_deploy/__main__.py [hunk 3]
```diff
@@ -92,16 +92,32 @@ def _join_image_registry(image, registry):
...
 
 
-def replace_tag_version(tag, version):
+def generate_versioned_tags(tag, version, latest=False):
     """
-    Replace the version information in the tag with the given version.
-
-    XXXX: will make more sense with more features...
+    Generate a list of tags based on the given tag and version information.
+    Prepends the given version
```
Comment (jerith) [inline docker_ci_deploy/__main__.py]: Is this docstring complete?

File: docker_ci_deploy/__main__.py [hunk 4]
```diff
@@ -222,12 +236,18 @@ def run(self, images, tags=None, version=None, login=None, registry=None):
...
+            Whether or not to tag this image with the latest tag. Requires a
+            version to be provided.
         :param login:
             Login details for the Docker registry in the form
             <username>:<password>.
         :param registry:
             The address to the Docker registry host.
         """
+        if latest and version is None:
+            raise ValueError('A version must be provided if latest is True')
```
Comment (jerith) [inline docker_ci_deploy/__main__.py]: Is `latest=True, version=''` a valid case?

File: docker_ci_deploy/tests/test_main.py [hunk 5]
```diff
@@ -380,6 +426,181 @@ def test_version_new_tag_with_version(self, capfd):
             'push test-image:1.2.3-def',
         ])
 
+    # FIXME?: The following 2 tests describe a weird, unintuitive edge case :-(
+    # Passing `--tag latest` with `--tag-version <version>` but *not*
+    # `--tag-latest` doesn't actually get you the tag 'latest' but rather
+    # effectively removes any existing tag.
```
Comment (jerith) [inline docker_ci_deploy/tests/test_main.py]: Maybe we should special-case this to set `--tag-latest` or something? That's for a new PR if we decide to do it, though.

File: docker_ci_deploy/__main__.py (hunk 4 above)
Comment (JayH5) [inline docker_ci_deploy/__main__.py]: Not *really*... good catch

File: docker_ci_deploy/tests/test_main.py (hunk 5 above)
Comment (JayH5) [inline docker_ci_deploy/tests/test_main.py]: Yeah I did consider special-casing it.. the logic just gets pretty messy then (and it's already not fantastic)
